// webpack-server/index.js
var express = require('express');
var path = require('path');
var _ = require('lodash');
var handlebars = require('handlebars');
var color = require('cli-color');
var fs = require('fs');
var url = require('url');

var defaults = {
  port: 8080,
  message: "\n\nserver listening at {{ url }}\n",
  index: 'index.html'
};

function createPlugin(opts) {
  var options = _.extend({}, defaults, opts);
  var template = handlebars.compile(options.message);

  return function serverPlugin() {
    if (!this.options.watch) return;

    var server = express()
      .use(liveStatics(options.root, options.index))
      .listen(options.port);

    this.plugin('after-compile', function startServer(err, done) {
      var message = template({ url: 'http://localhost:' + options.port });
      console.log(color.cyanBright(message));
      done();
    });
  };
}

module.exports = createPlugin;

function liveStatics(root, index) {
  if (!root) throw new Error('liveStatics() root path required');

  return function staticMiddleware(req, res, next) {
    if ('GET' != req.method && 'HEAD' != req.method) return next();

    var file = req.path === '/' ? '/' + index : req.path;
    var target = path.join(root, file);

    res.sendfile(target);
  };
}
